File: accounts/src/test/java/io/github/intisy/ai/shared/rank/LeaderboardTest.java
package io.github.intisy.ai.shared.rank;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class LeaderboardTest {

    private static List<Leaderboard.Score> scores(Object... pairs) {
        List<Leaderboard.Score> out = new ArrayList<>();
        for (int i = 0; i < pairs.length; i += 2) {
            out.add(new Leaderboard.Score((String) pairs[i], ((Number) pairs[i + 1]).doubleValue()));
        }
        return out;
    }

    @Test
    void normalizeDropsEffortTokensAndPunctuation() {
        assertEquals("claudeopus46", Leaderboard.normalize("claude-opus-4-6-thinking"));
        assertEquals("claude46opus", Leaderboard.normalize("Claude 4.6 Opus"));
    }

    @Test
    void everyVariantSpellingCollapsesToOneKey() {
        assertEquals("geminiflash", Leaderboard.normalize("Gemini Flash (High)"));
        assertEquals("geminiflash", Leaderboard.normalize("gemini-flash-high"));
        assertEquals("geminiflash", Leaderboard.normalize("gemini flash low"));
    }

    /** The provider label is parenthesised too, so only baseKeyFromName strips it. */
    @Test
    void baseKeyDropsEveryParentheticalTag() {
        assertEquals("claudeopusantigravity", Leaderboard.normalize("Claude Opus (High) (Antigravity)"));
        assertEquals("claudeopus", Leaderboard.baseKeyFromName("Claude Opus (High) (Antigravity)"));
    }

    @Test
    void effortRanksThinkingHighestAndPlainNamesMidRange() {
        assertEquals(6, Leaderboard.effortRank("Claude Opus (Thinking)"));
        assertEquals(5, Leaderboard.effortRank("Gemini Flash (High)"));
        assertEquals(4, Leaderboard.effortRank("Gemini Flash (Medium)"));
        assertEquals(3, Leaderboard.effortRank("Claude Opus"));
        assertEquals(2, Leaderboard.effortRank("gemini flash low"));
        assertEquals(1, Leaderboard.effortRank("Gemini Flash (Minimal)"));
    }

    /** Checked before plain "low", which would otherwise claim the same name at rank 2. */
    @Test
    void extraLowOutranksNothingAndIsNotReadAsLow() {
        assertEquals(1, Leaderboard.effortRank("Gemini Flash (Extra Low)"));
        assertEquals(1, Leaderboard.effortRank("gemini-flash-extra_low"));
    }

    @Test
    void scoreMatchesEitherWaySubstring() {
        assertEquals(50, Leaderboard.scoreForKey("geminiflash", scores("geminiflash", 50)), 1e-9);
        assertEquals(50, Leaderboard.scoreForKey("geminiflash", scores("geminiflashpro", 50)), 1e-9);
        assertEquals(50, Leaderboard.scoreForKey("geminiflashpro", scores("geminiflash", 50)), 1e-9);
    }

    @Test
    void takesTheBestOfSeveralMatches() {
        assertEquals(70, Leaderboard.scoreForKey("geminiflash", scores("geminiflash", 50, "geminiflashx", 70)), 1e-9);
    }

    /**
     * A catalog entry still ranks by its family when the score source lists only another version of
     * it, which is the whole point of the digit-stripped second pass.
     */
    @Test
    void fallsBackToTheDigitStrippedFamily() {
        assertEquals(70, Leaderboard.scoreForKey("claudeopus46", scores("claude48opus", 70)), 1e-9);
    }

    @Test
    void refusesAFamilyMatchOnAVeryShortKey() {
        assertEquals(Leaderboard.NO_SCORE, Leaderboard.scoreForKey("gpt", scores("claude48opus", 70)), 1e-9);
    }

    @Test
    void reportsNoScoreWithoutLiveData() {
        assertEquals(Leaderboard.NO_SCORE, Leaderboard.scoreForKey("claudeopus", scores()), 1e-9);
    }

    @Test
    void ordersByScoreBestFirst() {
        List<String> ids = Arrays.asList("a", "b", "c");
        List<String> names = Arrays.asList("Claude Opus", "Gemini Flash", "GPT Turbo");
        List<Leaderboard.Score> live = scores("claudeopus", 60, "geminiflash", 80, "gptturbo", 40);

        assertEquals(Arrays.asList("b", "a", "c"), Leaderboard.order(ids, names, live));
    }

    @Test
    void putsEveryScoredModelAheadOfEveryUnscoredOne() {
        List<String> ids = Arrays.asList("unscored", "scored");
        List<String> names = Arrays.asList("Totally Unknown Model", "Gemini Flash");

        assertEquals(Arrays.asList("scored", "unscored"),
                Leaderboard.order(ids, names, scores("geminiflash", 10)));
    }

    @Test
    void ordersVariantsOfOneModelByEffort() {
        List<String> ids = Arrays.asList("low", "thinking", "high");
        List<String> names = Arrays.asList("Gemini Flash (Low)", "Gemini Flash (Thinking)", "Gemini Flash (High)");

        assertEquals(Arrays.asList("thinking", "high", "low"),
                Leaderboard.order(ids, names, scores("geminiflash", 50)));
    }

    /** Effort is a within-model tie-break only: a low-effort better model still outranks a high-effort worse one. */
    @Test
    void effortNeverDecidesOrderBetweenDifferentModels() {
        List<String> ids = Arrays.asList("weakButThinking", "strongButLow");
        List<String> names = Arrays.asList("GPT Turbo (Thinking)", "Gemini Flash (Low)");

        assertEquals(Arrays.asList("strongButLow", "weakButThinking"),
                Leaderboard.order(ids, names, scores("gptturbo", 10, "geminiflash", 90)));
    }

    /**
     * The case that makes the naive comparator intransitive: two variants of one model separated in
     * the catalog by a different model. They must still come out grouped, and the sort must not
     * reject its own comparator.
     */
    @Test
    void groupsVariantsSeparatedInTheCatalogByAnotherModel() {
        List<String> ids = Arrays.asList("x-low", "y", "x-thinking");
        List<String> names = Arrays.asList("Gemini Flash (Low)", "GPT Turbo", "Gemini Flash (Thinking)");

        assertEquals(Arrays.asList("x-thinking", "x-low", "y"), Leaderboard.order(ids, names, scores()));
    }

    @Test
    void withoutLiveDataTheCatalogOrderOfDistinctModelsSurvives() {
        List<String> ids = Arrays.asList("first", "second", "third");
        List<String> names = Arrays.asList("Model A", "Model B", "Model C");

        assertEquals(ids, Leaderboard.order(ids, names, scores()));
    }

    /** A 40-entry catalog: enough that a merge sort would reject an intransitive comparator. */
    @Test
    void ordersACatalogLargeEnoughToTripTheMergeSortContractCheck() {
        List<String> ids = new ArrayList<>();
        List<String> names = new ArrayList<>();
        for (int i = 0; i < 40; i++) {
            ids.add("id" + i);
            names.add("Model " + (i % 4) + (i % 2 == 0 ? " (Thinking)" : " (Low)"));
        }

        List<String> ordered = Leaderboard.order(ids, names, scores());

        assertEquals(40, ordered.size());
        assertTrue(ordered.containsAll(ids));
    }

    @Test
    void scoresForCarriesOnlyTheIdsThatMatched() {
        Map<String, Double> expected = new LinkedHashMap<>();
        expected.put("known", Double.valueOf(50));

        assertEquals(expected, Leaderboard.scoresFor(
                Arrays.asList("known", "unknown"),
                Arrays.asList("Gemini Flash", "Totally Unknown Model"),
                scores("geminiflash", 50)));
    }

    @Test
    void idStandsInForAMissingName() {
        assertEquals(Arrays.asList("geminiflash"),
                Leaderboard.order(Arrays.asList("geminiflash"), null, scores("geminiflash", 50)));
    }

    @Test
    void sourceShortTagsLiveDataAndNothingElse() {
        assertEquals("AA", Leaderboard.sourceShort("Artificial Analysis via OpenRouter"));
        assertEquals("", Leaderboard.sourceShort(""));
        assertEquals("", Leaderboard.sourceShort(null));
    }
}

File: src/leaderboard.test.ts
import { afterEach, beforeEach, describe, expect, it } from "vitest";
import { mkdtempSync, mkdirSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

import { computeLeaderboardOrder, computeLeaderboardScores, leaderboardSourceShort } from "./leaderboard.js";

// A cache written inside the 24h TTL is what keeps these tests off the network: getScores() returns
// it without ever reaching for OpenRouter or Artificial Analysis.
function primeCache(home: string, scores: { norm: string; score: number }[]): void {
  mkdirSync(join(home, "config"), { recursive: true });
  writeFileSync(
    join(home, "config", "leaderboard.json"),
    JSON.stringify({ fetchedAt: Date.now(), source: "Artificial Analysis", scores }),
    "utf8",
  );
}

describe("leaderboard ranking", () => {
  let home = "";
  let previousHome: string | undefined;

  beforeEach(() => {
    home = mkdtempSync(join(tmpdir(), "leaderboard-"));
    previousHome = process.env.HUB_CONFIG_DIR;
    process.env.HUB_CONFIG_DIR = home;
  });

  afterEach(() => {
    if (previousHome === undefined) delete process.env.HUB_CONFIG_DIR;
    else process.env.HUB_CONFIG_DIR = previousHome;
    rmSync(home, { recursive: true, force: true });
  });

  it("orders a catalog best-first by live score", async () => {
    primeCache(home, [
      { norm: "claudeopus", score: 60 },
      { norm: "geminiflash", score: 80 },
      { norm: "gptturbo", score: 40 },
    ]);
    const names: Record<string, string> = { a: "Claude Opus", b: "Gemini Flash", c: "GPT Turbo" };

    expect(await computeLeaderboardOrder(["a", "b", "c"], (id) => names[id])).toEqual(["b", "a", "c"]);
  });

  it("puts every scored model ahead of every unscored one", async () => {
    primeCache(home, [{ norm: "geminiflash", score: 10 }]);
    const names: Record<string, string> = { known: "Gemini Flash", other: "Totally Unknown Model" };

    expect(await computeLeaderboardOrder(["other", "known"], (id) => names[id])).toEqual(["known", "other"]);
  });

  // Variants of one model separated in the catalog by a different model still come out grouped,
  // which the effort-versus-catalog-order tie-break cannot express on its own.
  it("groups variants of one model even when the catalog interleaves another", async () => {
    primeCache(home, []);
    const names: Record<string, string> = {
      "x-low": "Gemini Flash (Low)",
      y: "GPT Turbo",
      "x-thinking": "Gemini Flash (Thinking)",
    };

    expect(await computeLeaderboardOrder(["x-low", "y", "x-thinking"], (id) => names[id]))
      .toEqual(["x-thinking", "x-low", "y"]);
  });

  it("ranks variants of one model by effort", async () => {
    primeCache(home, [{ norm: "geminiflash", score: 50 }]);
    const names: Record<string, string> = {
      low: "Gemini Flash (Low)",
      thinking: "Gemini Flash (Thinking)",
      high: "Gemini Flash (High)",
    };

    expect(await computeLeaderboardOrder(["low", "thinking", "high"], (id) => names[id]))
      .toEqual(["thinking", "high", "low"]);
  });

  it("preserves the catalog order when there is no live data", async () => {
    primeCache(home, []);
    expect(await computeLeaderboardOrder(["first", "second", "third"], (id) => id))
      .toEqual(["first", "second", "third"]);
  });

  it("reports a score only for the ids that matched one", async () => {
    primeCache(home, [{ norm: "geminiflash", score: 50 }]);
    const names: Record<string, string> = { known: "Gemini Flash", other: "Totally Unknown Model" };

    expect(await computeLeaderboardScores(["known", "other"], (id) => names[id])).toEqual({ known: 50 });
  });

  // The score source lists a different version of the same family, which the digit-stripped
  // fallback still matches.
  it("ranks a catalog entry by its family when only another version is published", async () => {
    primeCache(home, [{ norm: "claude48opus", score: 70 }]);

    expect(await computeLeaderboardScores(["a"], () => "Claude 4.6 Opus")).toEqual({ a: 70 });
  });

  it("tags live data and nothing else", () => {
    expect(leaderboardSourceShort("Artificial Analysis via OpenRouter")).toBe("AA");
    expect(leaderboardSourceShort("")).toBe("");
  });
});
